File: modulo-java/src/com/fabrica/muebles/dao/ProveedorDAO.java
package com.fabrica.muebles.dao;

import com.fabrica.muebles.modelo.Proveedor;
import java.sql.*;
import java.util.*;
import com.fabrica.muebles.util.ConexionBD;

public class ProveedorDAO {

    // Usar la conexión unificada desde ConexionBD
    public Connection getConexion() {
        return ConexionBD.getConexion();
    }

    // Método para buscar un proveedor por ID
    public Proveedor buscar(int id) {
        String sql = "SELECT * FROM proveedor WHERE id = ?";
        Proveedor proveedor = null;

        try (Connection con = getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    proveedor = new Proveedor();
                    proveedor.setId(rs.getInt("id"));
                    proveedor.setNombre(rs.getString("nombre"));
                    proveedor.setContacto(rs.getString("contacto"));
                    proveedor.setTelefono(rs.getString("telefono"));
                    proveedor.setDireccion(rs.getString("direccion"));
                    proveedor.setCorreo(rs.getString("correo"));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return proveedor;
    }

    // Método para listar todos los proveedores
    public List<Proveedor> listar() {
        List<Proveedor> lista = new ArrayList<>();
        String sql = "SELECT * FROM proveedor";

        try (Connection con = getConexion();
             Statement st = con.createStatement();
             ResultSet rs = st.executeQuery(sql)) {

            while (rs.next()) {
                Proveedor p = new Proveedor();
                p.setId(rs.getInt("id"));
                p.setNombre(rs.getString("nombre"));
                p.setContacto(rs.getString("contacto"));
                p.setTelefono(rs.getString("telefono"));
                p.setDireccion(rs.getString("direccion"));
                p.setCorreo(rs.getString("correo"));
                lista.add(p);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return lista;
    }

    // Método para agregar un proveedor
    public boolean agregar(Proveedor p) {
        String sql = "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES (?, ?, ?, ?, ?)";

        try (Connection con = getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setString(1, p.getNombre());
            ps.setString(2, p.getContacto());
            ps.setString(3, p.getTelefono());
            ps.setString(4, p.getDireccion());
            ps.setString(5, p.getCorreo());
            ps.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // Método para actualizar un proveedor
    public boolean actualizar(Proveedor p) {
        String sql = "UPDATE proveedor SET nombre=?, contacto=?, telefono=?, direccion=?, correo=? WHERE id=?";

        try (Connection con = getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setString(1, p.getNombre());
            ps.setString(2, p.getContacto());
            ps.setString(3, p.getTelefono());
            ps.setString(4, p.getDireccion());
            ps.setString(5, p.getCorreo());
            ps.setInt(6, p.getId());
            ps.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // Método para eliminar un proveedor
    public boolean eliminar(int id) {
        String sql = "DELETE FROM proveedor WHERE id=?";

        try (Connection con = getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setInt(1, id);
            ps.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }
}

File: src/com/fabrica/muebles/dao/proveedor_dao.py
from com.fabrica.muebles.modelo.proveedor import Proveedor
from com.fabrica.muebles.util.conexion_bd import ConexionBD


class ProveedorDAO:
    """DAO para operaciones CRUD de Proveedores"""

    def buscar(self, id):
        """Busca un proveedor por ID"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute("SELECT * FROM proveedor WHERE id = %s", (id,))
            fila = cursor.fetchone()
            return Proveedor(*fila) if fila else None
        except Exception as e:
            print(f"Error: {e}")
            return None
        finally:
            cursor.close()
            conexion.close()

    def listar(self):
        """Lista todos los proveedores"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute("SELECT * FROM proveedor")
            return [Proveedor(*fila) for fila in cursor.fetchall()]
        except Exception as e:
            print(f"Error: {e}")
            return []
        finally:
            cursor.close()
            conexion.close()

    def agregar(self, p):
        """Agrega un nuevo proveedor"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES (%s, %s, %s, %s, %s)",
                (p.nombre, p.contacto, p.telefono, p.direccion, p.correo))
            conexion.commit()
            return True
        except Exception as e:
            print(f"Error: {e}")
            conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()

    def actualizar(self, p):
        """Actualiza un proveedor existente"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(
                "UPDATE proveedor SET nombre=%s, contacto=%s, telefono=%s, direccion=%s, correo=%s WHERE id=%s",
                (p.nombre, p.contacto, p.telefono, p.direccion, p.correo, p.id))
            conexion.commit()
            return True
        except Exception as e:
            print(f"Error: {e}")
            conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()

    def eliminar(self, id):
        """Elimina un proveedor"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute("DELETE FROM proveedor WHERE id=%s", (id,))
            conexion.commit()
            return True
        except Exception as e:
            print(f"Error: {e}")
            conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()